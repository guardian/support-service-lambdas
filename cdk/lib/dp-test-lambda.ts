import type { GuStackProps } from '@guardian/cdk/lib/constructs/core';
import { GuStack } from '@guardian/cdk/lib/constructs/core';
import { GuLambdaFunction } from '@guardian/cdk/lib/constructs/lambda';
import type { App } from 'aws-cdk-lib';
import { Effect, PolicyStatement } from 'aws-cdk-lib/aws-iam';
import { Architecture, LoggingFormat } from 'aws-cdk-lib/aws-lambda';
import { nodeVersion } from './node-version';

export class DpTestLambda extends GuStack {
	constructor(scope: App, id: string, props: GuStackProps) {
		super(scope, id, props);

		const appName = 'dp-test-lambda';

		const allowPutMetric = new PolicyStatement({
			effect: Effect.ALLOW,
			actions: ['cloudwatch:PutMetricData'],
			resources: ['*'],
		});

		new GuLambdaFunction(this, 'invoke-zuora-lambda', {
			app: appName,
			functionName: `${appName}-invoke-zuora-${this.stage}`,
			loggingFormat: LoggingFormat.TEXT,
			runtime: nodeVersion,
			environment: {
				Stage: this.stage,
			},
			handler: 'invokeZuora.handler',
			fileName: `${appName}.zip`,
			architecture: Architecture.ARM_64,
			initialPolicy: [
				allowPutMetric,
				new PolicyStatement({
					actions: ['secretsmanager:GetSecretValue'],
					resources: [
						`arn:aws:secretsmanager:${this.region}:${this.account}:secret:${this.stage}/Zuora-OAuth/SupportServiceLambdas-*`,
					],
				}),
			],
		});
	}
}
