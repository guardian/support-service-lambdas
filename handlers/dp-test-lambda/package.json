{
	"name": "dp-test-lambda",
	"description": "a lambda to test zuora rest response behaviour in CODE",
	"scripts": {
		"test": "jest --group=-integration",
		"it-test": "jest --group=integration",
		"build": "esbuild --bundle --platform=node --target=node20 --outdir=target src/handlers/*.ts",
		"lint": "eslint src/**/*.ts test/**/*.ts",
		"package": "pnpm type-check && pnpm lint && pnpm check-formatting && pnpm test && pnpm build && cd target && zip -qr negative-invoices-processor.zip ./*.js",
		"type-check": "tsc --noEmit",
		"check-formatting": "prettier --check **.ts",
		"fix-formatting": "prettier --write **.ts",
		"watch": "chokidar 'src/**/*.ts' -c 'pnpm fix-formatting'"
	},
	"devDependencies": {
		"@types/aws-lambda": "^8.10.147",
		"@types/aws-sdk": "^2.7.4"
	},
	"dependencies": {
		"aws-sdk": "^2.1692.0",
		"zod": "^3.23.8"
	}
}
