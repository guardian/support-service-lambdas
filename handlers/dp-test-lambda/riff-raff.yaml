stacks:
  - support
regions:
  - eu-west-1
allowedStages:
  - CODE
deployments:
  dp-test-lambda-cloudformation:
    type: cloud-formation
    app: dp-test-lambda
    parameters:
      templateStagePaths:
        CODE: dp-test-lambda-CODE.template.json
  dp-test-lambda:
    type: aws-lambda
    parameters:
      fileName: dp-test-lambda.zip
      bucketSsmLookup: true
      prefixStack: false
      functionNames:
        - dp-test-lambda-invoke-zuora-
    dependencies: [dp-test-lambda-cloudformation]
