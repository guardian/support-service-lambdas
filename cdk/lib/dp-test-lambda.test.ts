import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { DpTestLambda } from './dp-test-lambda';

describe('The dp-test-lambda stack', () => {
	it('matches the snapshot', () => {
		const app = new App();
		const codeStack = new DpTestLambda(app, 'dp-test-lambda-CODE', {
			stack: 'support',
			stage: 'CODE',
		});

		expect(Template.fromStack(codeStack).toJSON()).toMatchSnapshot();
	});
});
