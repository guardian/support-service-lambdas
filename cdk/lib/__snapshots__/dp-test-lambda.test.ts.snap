// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The dp-test-lambda stack matches the snapshot 1`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuDistributionBucketParameter",
      "GuLambdaFunction",
    ],
    "gu:cdk:version": "TEST",
  },
  "Parameters": {
    "DistributionBucketName": {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "invokezuoralambda17F17E77": {
      "DependsOn": [
        "invokezuoralambdaServiceRoleDefaultPolicyC9F1D70A",
        "invokezuoralambdaServiceRoleF5E15724",
      ],
      "Properties": {
        "Architectures": [
          "arm64",
        ],
        "Code": {
          "S3Bucket": {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "support/CODE/dp-test-lambda/dp-test-lambda.zip",
        },
        "Environment": {
          "Variables": {
            "APP": "dp-test-lambda",
            "STACK": "support",
            "STAGE": "CODE",
            "Stage": "CODE",
          },
        },
        "FunctionName": "dp-test-lambda-invoke-zuora-CODE",
        "Handler": "invokeZuora.handler",
        "LoggingConfig": {
          "LogFormat": "Text",
        },
        "MemorySize": 512,
        "Role": {
          "Fn::GetAtt": [
            "invokezuoralambdaServiceRoleF5E15724",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Tags": [
          {
            "Key": "App",
            "Value": "dp-test-lambda",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "support",
          },
          {
            "Key": "Stage",
            "Value": "CODE",
          },
        ],
        "Timeout": 30,
      },
      "Type": "AWS::Lambda::Function",
    },
    "invokezuoralambdaServiceRoleDefaultPolicyC9F1D70A": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "cloudwatch:PutMetricData",
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                      "/support/CODE/dp-test-lambda/dp-test-lambda.zip",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/CODE/support/dp-test-lambda",
                  ],
                ],
              },
            },
            {
              "Action": [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/CODE/support/dp-test-lambda/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "invokezuoralambdaServiceRoleDefaultPolicyC9F1D70A",
        "Roles": [
          {
            "Ref": "invokezuoralambdaServiceRoleF5E15724",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "invokezuoralambdaServiceRoleF5E15724": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "dp-test-lambda",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "support",
          },
          {
            "Key": "Stage",
            "Value": "CODE",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
